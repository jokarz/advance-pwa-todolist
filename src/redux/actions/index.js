//import axios from '../../utility/axiosInstance';
import { getData, setData } from '../../utility/localStorageHelper'
//-----------------------------View Archive---------------------$
export const setArchive = data => {
  return {
    type: 'setArchive',
    data
  }
}


//-----------------------------SUCCESS---------------------
export const resetSuccess = () => {
  return {
    type: 'resetSuccess'
  }
}

export const success = () => {
  return {
    type: 'success'
  }
}

export const successAt = data => {
  return {
    type: 'successAt',
    data
  }
}

//-----------------------LOADING---------------------------
export const loadStart = () => {
  return {
    type: 'loadStart'
  }
}

export const loadStop = () => {
  return {
    type: 'loadStop'
  }
}

//---------------------------------------ALERT----------------------
export const alertStart = () => {
  return {
    type: 'alertStart'
  }
}

export const alertStop = () => {
  return {
    type: 'alertStop'
  }
}

export const alertMsg = data => {
  return {
    type: 'updateAlertMsg',
    data
  }
}

export const alertAt = data => { //topic, todo+id, edit+id
  return {
    type: 'alertAt',
    data
  }
}


//-----------------------LOAD TOPIC AND TODOS--------------------------------
export const loadAll = data => {
  return {
    type: 'loadAll',
    data
  }
}


export const autoGetAll = () => {
  return async dispatch => {
    const data = getData()
    dispatch(loadAll(data))
  }
}

export const getAll = () => {
  return async dispatch => {
    const data = getData()
    dispatch(loadAll(data))
  }
}

export const addTopic = title => {
  return async dispatch => {
    title = title.trim()
    const data = getData()
    let index = data.findIndex(item => item.title === title)
    dispatch(loadStart());
    if (title === '') {
      dispatch(loadStop())
      dispatch(alertAt('topic'))
      dispatch(alertMsg('Please enter a valid topic'))
      dispatch(alertStart())
      setTimeout(() => { dispatch(alertStop()) }, 3000)
    } else if (index > -1) {
      dispatch(loadStop())
      dispatch(alertAt('topic'))
      dispatch(alertMsg(`Topic already exist ${data[index].archive ? 'in archived' : ''}`))
      dispatch(alertStart())
      setTimeout(() => { dispatch(alertStop()) }, 3000)
    } else {
      const id = Math.random().toString(36).substring(2)
      setData([{ id, title, archive: false, todos: { "completed": [], "incomplete": [] } }, ...data])
      dispatch(loadStop())
      dispatch(success())
      dispatch(successAt('topic' + title))
      dispatch(getAll())
    }
  }
}

export const archiveTopic = (topicId, archive) => {
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    dispatch(resetSuccess());
    let index = data.findIndex(item => item.id === topicId)
    if (index > -1) {
      data[index].archive = archive
      let tempo = data[index]
      data.splice(index, 1)
      data.unshift(tempo)
      if (!data.some(item => item.archive)) {
        dispatch(setArchive(false))
      }
    }
    setData(data)
    dispatch(loadStop())
    dispatch(successAt('topicarchive' + topicId))
    dispatch(success())
    dispatch(getAll())
    dispatch(resetSuccess());
  }
}

export const renameTopic = (topicId, title) => {
  return async dispatch => {
    const data = getData()
    let index = data.findIndex(item => item.title === title)
    dispatch(loadStart());
    dispatch(resetSuccess());
    if (index > -1) {
      dispatch(alertAt('topictitle' + topicId))
      dispatch(alertMsg(`Topic already exist ${data[index].archive ? 'in archived' : ''}`))
      dispatch(alertStart())
      setTimeout(() => { dispatch(alertStop()) }, 3000)
      dispatch(loadStop())
    } else {
      index = data.findIndex(item => item.id === topicId)
      if (index > -1) {
        data[index].title = title.trim()
      }
      setData(data)
      dispatch(getAll())
      dispatch(loadStop())
      dispatch(success())
      dispatch(successAt('topictitle' + topicId))
    }
    setTimeout(() => {
      dispatch(resetSuccess());
    }, 50)

  }
}

export const duplicateTopic = topicId => {
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    dispatch(resetSuccess())
    let dup = null
    for (let i = 0; i < data.length; i++) {
      let item = data[i]
      if (item.id === topicId) {
        dup = JSON.parse(JSON.stringify(item))
        break
      }
    }
    let index = 1
    while (true) {
      let exist = false
      let title = `(${index}) ${dup.title}`

      for (let i = 0; i < data.length; i++) {
        if (data[i].title === title) {
          exist = true
          break
        }
      }
      if (!exist) {
        dup.title = title
        dup.id = Math.random().toString(36).substring(2)
        data.unshift(dup)
        break;
      }
      index++
    }
    setData(data)
    dispatch(loadStop())
    dispatch(success())
    dispatch(successAt('topicduplicate' + topicId))
    dispatch(getAll())
    setTimeout(() => {
      dispatch(resetSuccess());
    }, 50)
  }
}

export const deleteTopic = topicId => {
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    let index = data.findIndex(item => item.id === topicId)
    if (index > -1) {
      data.splice(index, 1)
      if (!data.some(item => item.archive)) {
        dispatch(setArchive(false))
      }
      setData(data)
    }
    dispatch(loadStop())
    dispatch(success())
    dispatch(getAll())
  }
}

//this part onwards is done differently from the server sided ver
export const addTodo = (topicId, content) => {
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    dispatch(resetSuccess());
    let index = data.findIndex(item => item.id === topicId)
    if (index === -1) {
      dispatch(loadStop())
      return
    }
    const allTodos = [...data[index].todos.completed, ...data[index].todos.incomplete]
    if (allTodos.some(todo => todo.content === content.trim())) {
      dispatch(alertAt('todo' + topicId))
      dispatch(alertMsg('Todo already exist~'))
      dispatch(alertStart())
      setTimeout(() => { dispatch(alertStop()) }, 3000)
      dispatch(loadStop())
    } else {
      const newTodo = { id: Math.random().toString(36).substring(2), content }
      data[index].todos.incomplete.unshift(newTodo)
      setData(data)
      dispatch(loadStop())
      dispatch(successAt('addtodo' + topicId))
      dispatch(success())
      dispatch(getAll())
      setTimeout(() => {
        dispatch(resetSuccess());
      }, 50)
    }
  }
}

export const editTodos = (topicId, contents, action) => { // action = ['completed', 'incomplete', 'delete']
  return async dispatch => {
    const data = getData()
    dispatch(loadStart());
    dispatch(resetSuccess());
    let index = data.findIndex(item => item.id === topicId)
    if (index === -1) {
      dispatch(loadStop())
      return
    }

    if (action === 'incomplete' || action === 'completed') {
      let originType = action === 'incomplete' ? 'completed' : 'incomplete'
      contents.forEach(content => {
        let itemIndex = data[index]['todos'][originType].findIndex(item => item.content === content.trim())
        if (itemIndex === -1) {
          dispatch(loadStop())
          return
        }
        let item = data[index]['todos'][originType][itemIndex]
        data[index]['todos'][action].unshift({ ...item })
        data[index]['todos'][originType].splice(itemIndex, 1)
      });
    } else {
      contents.forEach(content => {
        const newIncomplete = data[index]['todos']['incomplete'].filter(todo => todo.content !== content.trim())
        const newCompleted = data[index]['todos']['completed'].filter(todo => todo.content !== content.trim())
        data[index]['todos']['incomplete'] = newIncomplete
        data[index]['todos']['completed'] = newCompleted
      })
    }
    setData(data)
    dispatch(loadStop())
    dispatch(successAt('edittodo' + topicId))
    dispatch(success())
    dispatch(getAll())
  }
}
