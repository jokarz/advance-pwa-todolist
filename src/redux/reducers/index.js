import { combineReducers } from 'redux'
import Topic from './topicReducer.js'
import Misc from './miscReducer.js'
export default combineReducers({
  Misc, Topic
})
