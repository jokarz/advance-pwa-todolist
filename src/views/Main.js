import React, { useState, useEffect } from 'react'
import Masonry from 'react-masonry-css'
import { css } from 'aphrodite'
import { connect } from 'react-redux'

import Card from '../components/card/Card'
import Fab from '../components/fab/Fab'
import TopicModal from '../components/modal/TopicModal'
import { getAll, setArchive } from '../redux/actions'
import styles from '../utility/styles'

const Main = props => {
  const { Topic, Misc, getAll, setArchive } = props
  const [showModal, setShowModal] = useState(false)
  const handleModalClose = () => setShowModal(false)
  const [hideFab, setHideFab] = useState(false)

  useEffect(() => {
    getAll()
    // eslint-disable-next-line
  }, [])

  return (
    <>
      {Topic != null && Topic.length === 0
        ? <div className='d-flex justify-content-center align-items-center height-90'>
          <h5 className={`text-center ${css(styles.fontTitan)}`}>
            Create new topic to get started
          </h5>
          </div>
        : <>
          <div className='row'>
            <div className={`col d-flex justify-content-center fixed-top ${css(styles.eventPassThrough)}`}>
              <h4
                onClick={() => setArchive(!Misc.viewArchive)}
                className={`mt-5 text-center p-3 select-none border-secondary bg-white ${css(styles.pointer, styles.radius, styles.fontTitanSmaller, styles.shadowStatic, styles.borderThicker, styles.fabSubButtonHover, styles.eventAuto, styles.bounceInFast)}`}
              >
                {Misc.viewArchive ? 'Archived' : 'Ongoing'}
                {/* {Topic != null && Topic.filter(item => Boolean(item.archive) === Misc.viewArchive).length > 1 ? 's' : ''} */}
                <span className='fas fa-sort ml-2' />
              </h4>
            </div>
          </div>
          {
            Topic != null && Topic.filter(item => Boolean(item.archive) === Misc.viewArchive).length > 0
              ? <Masonry breakpointCols={{ default: 3, 950: 3, 800: 2, 650: 1 }} className='mg pb-5' columnClassName='mg_column'>
                {Topic.filter(item => Boolean(item.archive) === Misc.viewArchive).map((item, i) => (
                  <Card
                    entrance={i * 0.001}
                    id={item.id}
                    key={item.id}
                    completed={item.todos.completed || []}
                    incomplete={item.todos.incomplete || []}
                    title={item.title}
                    setHideFab={setHideFab}
                  />
                ))}
                </Masonry>
              : Topic != null
                ? <div className={`d-flex justify-content-center align-items-center ${css(styles.altHeight)}`}>
                  <h2 className={`text-center ${css(styles.fontTitan)}`}>
                    {Misc.viewArchive
                      ? <span>
                        No archived topic found
                        </span>
                      : <span>
                        Click below to create a new topic
                        </span>}
                  </h2>
                  </div>
                : null
          }
        </>}
      {
        hideFab && window.innerWidth < 600 ? null
          : <Fab
            setShowTopicModal={setShowModal}
            setArchive={setArchive}
            archive={Misc.viewArchive}
            archiveTopicCount={Topic != null ? Topic.filter(item => Boolean(item.archive)).length : 0}
            />
      }

      <TopicModal show={showModal} closeOption={handleModalClose} />
    </>
  )
}

const mapStateToProps = state => ({
  ...state
})

const mapDispatchToProps = dispatch => ({
  getAll: () => dispatch(getAll()),
  setArchive: value => dispatch(setArchive(value))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Main)
