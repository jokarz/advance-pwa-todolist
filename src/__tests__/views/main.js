import React from 'react'
import Main from '../../views/Main'
import { Provider } from 'react-redux'
import { createStore, applyMiddleware } from 'redux'
import thunkMiddleware from 'redux-thunk'

import rootReducer from '../../redux/reducers'

import { render, fireEvent } from '@testing-library/react'

test('Default setup renders', async () => {
  const store = createStore(
    rootReducer,
    applyMiddleware(
      thunkMiddleware
    )
  )
  const { asFragment} = render(<Provider store={store}>
    <Main />
  </Provider>)
  expect(asFragment()).toMatchSnapshot();
})

test('Modal renders', async () => {
  const store = createStore(
    rootReducer,
    applyMiddleware(
      thunkMiddleware
    )
  )
  const { container} = render(<Provider store={store}>
    <Main />
  </Provider>)
  fireEvent.click(container.querySelector('.fas.fa-plus'))
  expect(container.querySelector('.modal.d-block')).toBeTruthy()
})
