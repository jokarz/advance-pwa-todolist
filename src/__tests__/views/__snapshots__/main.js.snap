// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default setup renders 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center align-items-center height-90"
  >
    <h5
      class="text-center fontTitan_dzn0bb"
    >
      Create new topic to get started
    </h5>
  </div>
  <div
    class="fixed-bottom text-center mb-5 eventPassThrough_1dwq0yf-o_O-bounceIn_av56gd"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <button
          class="select-none btn shadow btn-lg rounded-circle btn-primary
              fabSubButtonHover_wu4qfj-o_O-eventAuto_ow2iz6-o_O-bounceInFast_tj83ey-o_O-fontTitanSmaller_g467oi
              "
          type="button"
        >
          <span
            class="fas fa-plus"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="modal d-none"
  >
    <div
      class="backDrop_4pfu4w-o_O-modalBackground_122ndvs"
    />
    <div
      class="modal-dialog mt-4 bounceOutUp_wlq44y"
    >
      <div
        class="modal-content radius_1fb25e"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title text-center m-auto fontTitanSmaller_g467oi"
          >
            New Topic
          </h5>
        </div>
        <div
          class="modal-body"
        >
          <input
            class="form-control  "
            placeholder="What new topic you have in mind~"
            type="text"
            value=""
          />
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-secondary radius_1fb25e-o_O-shadow_jlwb8p"
            type="button"
          >
            <span
              class="mr-1"
            >
              Nahh
            </span>
            <span
              class="fas fa-times"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
